### PR TITLE
Unstable opts: object spread/rest

### PR DESCRIPTION
**Pull request by jugglinmike**
As the first "unstable" option, this changeset includes a commit that implements the internals necessary to support options of that type (resolving gh-2601). The second commit introduces the first unstable option--experimental support for [the object spread/rest](https://github.com/tc39/proposal-object-rest-spread) proposal (resolving gh-2991).

### CODE REVIEW

File: src/lex.js [hunk 1]
```diff
@@ -405,7 +405,10 @@ Lexer.prototype = {
     // comments.
 
     function commentToken(label, body, opt) {
-      var special = ["jshint", "jslint", "members", "member", "globals", "global", "exported"];
+      var special = [
+        "jshint", "jshint.unstable", "jslint", "members", "member", "globals",
```
Comment (rwaldron) [inline src/lex.js]: It's like _MetaProperty_

File: tests/unit/options.js [hunk 2]
```diff
@@ -4094,3 +4094,41 @@ exports.trailingcomma = function (test) {
...
+    .test("", { unstable: { undef: true } });
+
+  TestRun(test, "Rejects stable names specified via in-line directive (single-line comment).")
+    .addError(1, "Bad unstable option: 'undef'.")
+    .test("// jshint.unstable undef: true");
+
+  TestRun(test, "Rejects stable names specified via in-line directive (multi-line comment).")
+    .addError(1, "Bad unstable option: 'undef'.")
+    .test("/* jshint.unstable undef: true */");
+
```
Comment (rwaldron) [inline tests/unit/options.js]: Thorough!

File: tests/unit/unstable/objrestspread.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    .addError(4, "Invalid element after rest element.")
+    .addError(5, "Invalid element after rest element.")
+    .test(code, { esversion: 6, unstable: { objspreadrest: true } });
+
+  code = [
+    "({ ...[a, b, c] } = {});",
+    "({ a, ...[b, c, d] } = {});",
+    "({ a = 0, ...[b, c, d] } = {});",
+    "({ a: A, ...[b, c, d] } = {});",
+    "({ a: A = 0, ...[b, c, d] } = {});"
```
Comment (rwaldron) [inline tests/unit/unstable/objrestspread.js]: Per recent consensus, this is invalid: `{ ...[] } = {}`

- http://tc39.github.io/tc39-notes/2017-05_may-23.html#16ih-why-allow-bindingpattern-for-bindingrestparameter-for-object-rest-maybe-we-should-just-allow-identifiers
- http://tc39.github.io/tc39-notes/2017-05_may-23.html#conclusionresolution-16ih-why-allow-bindingpattern-for-bindingrestparameter-for-object-rest-maybe-we-should-just-allow-identifiers

File: tests/unit/unstable/objrestspread.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    .addError(5, "Invalid element after rest element.")
+    .test(code, { esversion: 6, unstable: { objspreadrest: true } });
+
+  TestRun(test, "nested array pattern, empty")
+    .addError(1, "Empty destructuring.")
+    .test("({ ...[] } = {});", { esversion: 6, unstable: { objspreadrest: true } });
+
+  TestRun(test, "nested object pattern, empty")
+    .addError(1, "Empty destructuring.")
+    .test("({ ...{} } = {});", { esversion: 6, unstable: { objspreadrest: true } });
```
Comment (rwaldron) [inline tests/unit/unstable/objrestspread.js]: The consensus referenced above applies to `{ ...{} } = {}` as well

**Comment by rwaldron** [comment]
Nice work @jugglinmike!

**Comment by vernak2539** [comment]
Thanks for taking care of this! Any idea when this will be released?

**Pull request merged**